src/hazelcore/semantics/TextShape.re

```diff
@@ -9,34 +9,46 @@ type t =
   | ExpandingKeyword(ExpandingKeyword.t)
   | Var(Var.t);
 
+let int_of_intlit_opt = (n: string): option(int) => {
+  let stripped_text = StringUtil.strip_underscore(n);
+  int_of_string_opt(stripped_text);
+};
+
+let float_of_floatlit_opt = (f: string): option(float) => {
+  let stripped_text = StringUtil.strip_underscore(f);
+  float_of_string_opt(stripped_text);
+};
+
 let of_text = (text: string): option(t) => {
+  let stripped_text = StringUtil.strip_underscore(text);
   switch (
-    int_of_string_opt(text),
-    float_of_string_opt(text),
+    int_of_string_opt(stripped_text),
+    float_of_string_opt(stripped_text),
     bool_of_string_opt(text),
     ExpandingKeyword.mk(text),
   ) {
   /* the num_digits double count with leading zeros when the int is evaluated to 0  */
   | (Some(n), _, _, _) when n == 0 =>
-    StringUtil.num_leading_zeros(text) == String.length(text)
+    StringUtil.num_leading_zeros(stripped_text)
+    == String.length(stripped_text)
       ? Some(IntLit(text)) : None
   | (Some(n), _, _, _) =>
-    // OCaml accepts and ignores underscores
-    // when parsing ints and floats from strings, we don't
-    StringUtil.num_leading_zeros(text)
-    + IntUtil.num_digits(n) == String.length(text)
+    StringUtil.num_leading_zeros(stripped_text)
+    + IntUtil.num_digits(n) == String.length(stripped_text)
       ? Some(IntLit(text)) : None
   /* 1 is subtracted from num_digits because Ocaml introduces extra 0 in front of the decimal when float is < 1 */
   | (_, Some(f), _, _) when Float.abs(f) < 1.0 =>
-    StringUtil.num_leading_zeros(text)
+    StringUtil.num_leading_zeros(stripped_text)
```

Comment:
same here, why are we special casing these things? can you provide examples of text that goes into each branch in comments?